# TP Noté Informatique Quantique (Polytech Paris-Saclay 2024/2025)

Consignes:
- Tout ce dont vous avez besoin pour programmer durant cet examen est importé dans les cellules ci-dessous.
- Vous avez le droit aux documents, notamment les sujets et corrigés des TP (voir archive).
- C'est un examen **INDIVIDUEL**, vous ne pouvez donc pas communiquer avec vos camarades. Deux copies identiques seront sanctionnées sans discernement.
- L'examen est noté sur 20 points. Le détail des points est précisé au niveau du titre de chaque exercice.
- A la fin du TP, envoyer ce notebook en le renommant "examen_TP_prenom_nom.ipynb" à l'adresse mail : oceanekoska@gmail.com

**Executez cette cellule et les suivantes avant de commencer le TP**

$$\renewcommand{\ket}[1]{\left|{#1}\right\rangle}$$
$$\renewcommand{\bra}[1]{\left\langle{#1}\right|}$$

In [68]:
import qiskit
import qiskit_aer
import numpy as np

from qiskit_aer import Aer, AerSimulator
from qiskit import QuantumCircuit, assemble, QuantumRegister, ClassicalRegister, transpile
from qiskit.circuit.library import RYGate, RZGate, RXGate, TGate, TdgGate, HGate, MCXGate
from qiskit.visualization import plot_histogram, array_to_latex

Fonctions pour la simulation de circuit

In [69]:
# Retourne l'état de sortie
def simulate_circuit_state(qc):
    qc2 = qc.copy()
    aer_sim = Aer.get_backend('aer_simulator')
    qc2.save_statevector()
    result = aer_sim.run(qc2).result()
    
    return result.get_statevector().data

# Retourne l'unitaire complet U
def simulate_circuit_unitary(qc):
    qc2 = qc.copy()
    aer_sim = Aer.get_backend('aer_simulator')
    qc2.save_unitary()
    result = aer_sim.run(qc2).result()
        
    return result.get_unitary().data

# Retourne le résultat de la mesure sur les qubits donnés en input
def simulate_circuit_measurements(qc, qubits):
    qc2 = qc.copy()
    for (j,i) in enumerate(qubits):
        qc2.measure(i,j)
    
    aer_sim = Aer.get_backend('aer_simulator')
    result = aer_sim.run(qc2, shots=1, memory=True).result()
    
    return result.get_memory()

# Retourne l'histogramme des résultats : un dictionnaire où chaque clé est une bitstring et la valeur
# associée est le nombre de fois que cette bitstring a été mesurée. Le nombre de samples est 1024 par défaut.
def simulate_circuit_histogram(qc, qubits):
    qc2 = transpile(qc, AerSimulator())
    for (j,i) in enumerate(qubits):
        qc2.measure(i,j)
    aer_sim = Aer.get_backend('aer_simulator')
    result = aer_sim.run(qc2).result()
    
    return result.get_counts()

# Retourne le dessin
def simulate_circuit_drawing(qc):
    A = simulate_circuit_unitary(qc)
    B = np.where(A != 0, "#", ".")
    for l in B:
        for c in l:
            print(c, end="  ")
        print("\n")

---
# Début du sujet

Aujourd'hui vous allez être parent, parent d'un petit ordinateur quantique. Il s'appelle <span style="color:aquamarine"> mégatron (1 point bonus sur la note gratuit)</span>
  et il rentre en maternelle. Vous allez devoir l'aider à faire ses devoirs tout au long de sa scolarité pour qu'il devienne un bon élève. 

---
## Exercice 1 : Faire des dessins (2.5 points)

Dessiner est une tâche qu'on apprend très tôt. Pour un ordinateur quantique c'est pas trivial mais on va adapter un peu. Un circuit quantique représente une unitaire, donc une matrice. La matrice sera donc la toile de dessin. Chaque case de la matrice peut contenir 0 ou une valeurs non nulle. Si on considère :
- 0 = pixel blanc
- pas 0 = pixel noir

On peut faire du pixel art en noir et blanc ! Dans la suite, vous allez implémenter deux circuits quantiques et grace à la fonction simulate_circuit_drawing pouvoir afficher le dessin.

---
### Une croix

On commence par faire simple. Dessiner une croix $\times$ c'est en somme deux traits, plutôt simple quoi. Voici le circuit quantique correspondant, pour $n=3$ qubits :

<center>
    <img src="img/circuit_croix.png" />
</center>

#### **Question 1.1** 
*Implémenter le circuit quantique ci-dessus et afficher le dessin associé à l'aide de la fonction simulate_circuit_drawing*

In [70]:
qc = QuantumCircuit(3)
qc.cx(0,2)
qc.cx(0,1)
qc.h(0)
qc.cx(0,1)
qc.cx(0,2)
simulate_circuit_drawing(qc)

#  .  .  .  .  .  .  #  

.  #  .  .  .  .  #  .  

.  .  #  .  .  #  .  .  

.  .  .  #  #  .  .  .  

.  .  .  #  #  .  .  .  

.  .  #  .  .  #  .  .  

.  #  .  .  .  .  #  .  

#  .  .  .  .  .  .  #  



---
### Oeil de Sauron

Que l'on sache ou non ce qu'est un oeil de Sauron, il faut juste savoir que c'est plus compliqué qu'une croix et que ça ressemble à ceci

<center>
    <img src="img/os.jpg" />
</center>

Voici le circuit correspondant pour faire le dessin pour $n=4$ qubits:

<center>
    <img src = "img/circuit_sauron.png" />
</center>

#### **Question 1.2**

*Implémenter le circuit quantique ci-dessus et afficher le dessin associé à l'aide de la fonction simulate_circuit_drawing*


In [71]:
qc = QuantumCircuit(4)
qc.x(3)
qc.cx(3,0)
qc.x(0)
qc.cx(3,1)
qc.cx(0,1)
qc.cx(3,2)
qc.ccx(0,1,2)
qc.x(3)
qc.h(3)
qc.cx(3,0)
qc.cx(3,1)
qc.cx(3,2)
simulate_circuit_drawing(qc)

.  .  .  .  .  .  #  .  .  #  .  .  .  .  .  .  

.  .  .  .  .  #  .  .  .  .  #  .  .  .  .  .  

.  .  .  .  #  .  .  .  .  .  .  #  .  .  .  .  

.  .  .  #  .  .  .  .  .  .  .  .  #  .  .  .  

.  .  #  .  .  .  .  .  .  .  .  .  .  #  .  .  

.  #  .  .  .  .  .  .  .  .  .  .  .  .  #  .  

#  .  .  .  .  .  .  .  .  .  .  .  .  .  .  #  

.  .  .  .  .  .  .  #  #  .  .  .  .  .  .  .  

.  .  .  .  .  .  .  #  #  .  .  .  .  .  .  .  

#  .  .  .  .  .  .  .  .  .  .  .  .  .  .  #  

.  #  .  .  .  .  .  .  .  .  .  .  .  .  #  .  

.  .  #  .  .  .  .  .  .  .  .  .  .  #  .  .  

.  .  .  #  .  .  .  .  .  .  .  .  #  .  .  .  

.  .  .  .  #  .  .  .  .  .  .  #  .  .  .  .  

.  .  .  .  .  #  .  .  .  .  #  .  .  .  .  .  

.  .  .  .  .  .  #  .  .  #  .  .  .  .  .  .  



---
## Exercice 2 : Addition avec retenue (7.5 points)

Votre enfant ordinateur quantique est tellement rapide qu'il a sauté le CP et rentre directement au CE1. Maintenant, il va passer aux choses sérieuses et faire des additions **AVEC** retenue. Il a le droit a un traitement particulier : il peut faire ses additions en binaire. Nous allons donc étudier le problème de l'additionneur quantique. Considérons deux entiers encodés sur $3$ bits 
$$
a = a_2 a_1 a_0 \; \textrm{et} \; b = b_2 b_1 b_0
$$
Nous souhaitons calculer la somme exacte de ces deux entiers et stocker cette valeur, qui sera donc encodée sur 4 bits pour prendre en considération une potentielle retenue. Pour ce faire, on va avoir besoin de deux opérations que nous allons nommer $O_1$ et $O_2$ par manque d'originalité.

---
### Opérateur $O_1$

L'opérateur $O_1$ est un opérateur sur 3 qubits ayant pour circuit quantique

<center>
    <img src="img/circuit_O1.png" />
</center>

#### **Question 2.1**
*Implémenter l'opérateur O1 à l'aide du circuit quantique donné ci-dessus*

In [72]:
o1 = QuantumCircuit(3)
o1.cx(2,1)
o1.cx(2,0)
o1.ccx(0,1,2)
o1.draw()

┌───┐     
q_0: ─────┤ X ├──■──
     ┌───┐└─┬─┘  │  
q_1: ┤ X ├──┼────■──
     └─┬─┘  │  ┌─┴─┐
q_2: ──■────■──┤ X ├
               └───┘

#### **Question 2.2**
Supposons que le système en entrée de $O_1$ soit $\ket{a_i} \ket{b_i} \ket{c_i}$

*Expliciter l'état en sortie à la main (en utilisant l'opérateur xor (^), correspondant au "ou" exclusif)*

Votre réponse ici :
$\ket{a_i xor c_i} \ket{b_i xor c_i} \ket{c_i xor (b_i xor c_i) and (a_i xor c_i)}$

*Vérifier numériquement, en affichant pour chaque entrée possible sur 3 qubits, votre réponse*

In [73]:
n = o1.num_qubits

# Pour avoir un bel affichage
a = "{:^" + str(n + 3) + "s}|{:^" + str(n + 3) + "s}|{:^" + str(n + 3) + "s}"
print(a.format("Input", "Output", "Expected Output"))
print(a.format("", "", ""))
    
# L'itération sur les bitstrings
for i in range(2**n):
    circ = QuantumCircuit(n, n)
    
    # On récupère la décomposition binaire...
    bin_i = ("{0:0" + str(n) + "b}").format(i)
    
    # ... qui nous donne les X à appliquer
    for j in range(n):
        if bin_i[j] == '1':
            circ.x(j)
        
    # La méthode compose permet de concaténer deux circuits
    circ = circ.compose(o1, [i for i in range(n)])
    
    # On effectue la mesure
    output = simulate_circuit_measurements(circ, [i for i in range(n)])[0][::-1]
    
    # Calcul de la sortie attendue (basée sur l'opération XOR)
    a_i, b_i, c_i = map(int, bin_i)  # On décompose l'entrée
    expected_output = f"{(a_i ^ c_i)}{(b_i ^ c_i)}{(c_i ^ ((b_i ^ c_i) & (a_i ^ c_i)))}"
    
    # On affiche le résultat avec la sortie attendue
    print(a.format(bin_i, output, expected_output))


Input |Output|Expected Output
      |      |      
 000  | 000  | 000  
 001  | 110  | 110  
 010  | 010  | 010  
 011  | 101  | 101  
 100  | 100  | 100  
 101  | 011  | 011  
 110  | 111  | 111  
 111  | 001  | 001  


---

### Opérateur $O_2$

L'opérateur $O_2$ est aussi un opérateur sur 3 qubits, dont voici le circuit

<center>
    <img src="img/circuit_O2.png" />
</center>

#### **Question 2.3**
*Implémenter l'opérateur O2 à partir du circuit quantique ci-dessus*

In [74]:
o2 = QuantumCircuit(3)
o2.ccx(0,1,2)
o2.cx(2,0)
o2.cx(0,1)
o2.draw()

┌───┐     
q_0: ──■──┤ X ├──■──
       │  └─┬─┘┌─┴─┐
q_1: ──■────┼──┤ X ├
     ┌─┴─┐  │  └───┘
q_2: ┤ X ├──■───────
     └───┘

---
### Additionneur

Maintenant grâce à nos deux opérateurs $O_1$ et $O_2$, notre petit ordinateur quantique peut en théorie faire des additions avec retenue. Il suffit de faire une combinaison astucieuse de ces deux opérateurs pour obtenir le résultat. Toujours avec nos entiers  $a= a_2 a_1 a_0$ et $b = b_2 b_1 b_0$, nous allons considérer un vecteur $c$ sur 4 qubits. Ce vecteur $c$ va nous permettre de compter les retenues pendant l'addition. On a alors 
- $c_0 = 0$, car on commence quoi qu'il arrive le calcul sans retenue,
- $c_{i+1} = a_i \oplus b_i \oplus c_i$, où $\oplus$ est un ou exclusif (xor).


#### **Question 2.4**
*Vérifier sur toutes les entrées possibles de ce circuit qu'en sortie c_i et a_i soient inchangées.*
<center>
    <img src="img/circuit_O1O2.png" />
</center>


In [75]:
combined_circuit = QuantumCircuit(3)
combined_circuit.compose(o1, [0, 1, 2], inplace=True)
combined_circuit.compose(o2, [0, 1, 2], inplace=True)
combined_circuit.draw()

┌───┐          ┌───┐     
q_0: ─────┤ X ├──■────■──┤ X ├──■──
     ┌───┐└─┬─┘  │    │  └─┬─┘┌─┴─┐
q_1: ┤ X ├──┼────■────■────┼──┤ X ├
     └─┬─┘  │  ┌─┴─┐┌─┴─┐  │  └───┘
q_2: ──■────■──┤ X ├┤ X ├──■───────
               └───┘└───┘

In [76]:
n = combined_circuit.num_qubits

# Pour avoir un bel affichage
a ="{:^"+str(n+3)+"s}|{:^"+str(n+3)+"s}"
print(a.format("Input", " Output"))
print(a.format("", ""))
    
# L'itération sur les bitstrings
for i in range(2**n):
    circ = QuantumCircuit(n, n)
    
    # On récupère la décomposition binaire...
    bin_i = ("{0:0" + str(n) + "b}").format(i)
    
    # ... qui nous donne les X à appliquer
    for j in range(n):
        if bin_i[j] == '1':
            circ.x(j)
        
    # La méthode compose permet de concaténer deux circuits
    circ = circ.compose(combined_circuit, [i for i in range(n)])
    
    # On effectue la mesure
    output = simulate_circuit_measurements(circ, [i for i in range(n)])[0][::-1]
    
    # On affiche le résultat avec la sortie attendue
    print(a.format(bin_i, output, expected_output))

Input | Output
      |      
 000  | 000  
 001  | 011  
 010  | 010  
 011  | 001  
 100  | 110  
 101  | 101  
 110  | 100  
 111  | 111  


*Quelle est la valeur du qubit initialement à b_i (en fonction a_i, b_i, c_i) ?*

\ket{(b_i xor c_i) xor ((a_i xor c_i) xor ((c_i xor (b_i xor c_i) and (a_i xor c_i)) xor (a_i xor c_i) & (b_i xor c_i)))}

Il est temps de construire l'additionneur au complet pour des entiers sur 3 qubits. Pour ne pas faire durer plus longtemps le suspens, voici le circuit quantique

<center>
    <img src="img/circuit_adder.png" />
</center>

#### **Question 2.5**
*Implémenter le circuit de l'additionneur donné ci-dessus*

In [77]:
addition = QuantumCircuit(8)
addition.compose(o1, [0, 1, 2], inplace=True)
addition.compose(o1, [2, 3, 4], inplace=True)
addition.compose(o1, [4, 5, 6], inplace=True)
addition.cx(6, 7)
addition.compose(o2, [4, 5, 6], inplace=True)
addition.compose(o2, [2, 3, 4], inplace=True)
addition.compose(o2, [0, 1, 2], inplace=True)
addition.draw()

┌───┐                                                            »
q_0: ─────┤ X ├──■─────────────────────────────────────────────────────────»
     ┌───┐└─┬─┘  │                                                         »
q_1: ┤ X ├──┼────■─────────────────────────────────────────────────────────»
     └─┬─┘  │  ┌─┴─┐┌───┐                                        ┌───┐     »
q_2: ──■────■──┤ X ├┤ X ├──■──────────────────────────────────■──┤ X ├──■──»
     ┌───┐     └───┘└─┬─┘  │                                  │  └─┬─┘┌─┴─┐»
q_3: ┤ X ├────────────┼────■──────────────────────────────────■────┼──┤ X ├»
     └─┬─┘            │  ┌─┴─┐┌───┐               ┌───┐     ┌─┴─┐  │  └───┘»
q_4: ──■──────────────■──┤ X ├┤ X ├──■─────────■──┤ X ├──■──┤ X ├──■───────»
     ┌───┐               └───┘└─┬─┘  │         │  └─┬─┘┌─┴─┐└───┘          »
q_5: ┤ X ├──────────────────────┼────■─────────■────┼──┤ X ├───────────────»
     └─┬─┘                      │  ┌─┴─┐     ┌─┴─┐  │  └───┘               »
q_6: ──■────────────────────────■──┤ X ├──■──┤ X ├──■──────────────────────»
                                   └───┘┌─┴─┐└───┘                         »
q_7: ───────────────────────────────────┤ X ├──────────────────────────────»
                                        └───┘                              »
«          ┌───┐     
«q_0: ──■──┤ X ├──■──
«       │  └─┬─┘┌─┴─┐
«q_1: ──■────┼──┤ X ├
«     ┌─┴─┐  │  └───┘
«q_2: ┤ X ├──■───────
«     └───┘          
«q_3: ───────────────
«                    
«q_4: ───────────────
«                    
«q_5: ───────────────
«                    
«q_6: ───────────────
«                    
«q_7: ───────────────
«

#### **Question 2.6**
*Combien font 2 + 4 ? (C'est à l'ordinateur quantique de répondre of course)*

In [ ]:
# a = 0b010
# b = 0b100
bin_in = ""


n = addition.num_qubits

# Pour avoir un bel affichage
a ="{:^"+str(n+3)+"s}|{:^"+str(n+3)+"s}"
print(a.format("Input", " Output"))
print(a.format("", ""))
    
circ = QuantumCircuit(n, n)

# car c0=0 b0 a0 b1 a1 b2 a2 c3=0
bin_i = "01001000"

# ... qui nous donne les X à appliquer
for j in range(n):
    if bin_i[j] == '1':
        circ.x(j)
    
# La méthode compose permet de concaténer deux circuits
circ = circ.compose(addition, [i for i in range(n)])

# On effectue la mesure
output = simulate_circuit_measurements(circ, [i for i in range(n)])[0][::-1]

# On affiche le résultat avec la sortie attendue
print(a.format(bin_i, output, expected_output))

# res : 01011000  donc 1100 donc 110 donc 6

   Input   |   Output  
           |           
 01001000  | 01011000  


*Combien font 5 + 7 ? (Idem)*

In [ ]:
# a = 0b101
# b = 0b111
bin_in = ""


n = addition.num_qubits

# Pour avoir un bel affichage
a ="{:^"+str(n+3)+"s}|{:^"+str(n+3)+"s}"
print(a.format("Input", " Output"))
print(a.format("", ""))
    
circ = QuantumCircuit(n, n)

# car c0=0 b0 a0 b1 a1 b2 a2 c3=0
bin_i = "01110110"

# ... qui nous donne les X à appliquer
for j in range(n):
    if bin_i[j] == '1':
        circ.x(j)
    
# La méthode compose permet de concaténer deux circuits
circ = circ.compose(addition, [i for i in range(n)])

# On effectue la mesure
output = simulate_circuit_measurements(circ, [i for i in range(n)])[0][::-1]

# On affiche le résultat avec la sortie attendue
print(a.format(bin_i, output, expected_output))

# res : 00100111  donc 1100 donc 12

   Input   |   Output  
           |           
 01110110  | 00100111  


---
## Exercice 3 : Message caché (10 points)

L'école ce n'est pas seulement travailler, c'est aussi faire des rencontres. Votre petit ordinateur quantique a de nouveaux amis dans sa classe de CE2, Arthur (A) et Béatrice (B), et il aimerait les aider à communiquer secrétement en cours. Seulement, il ne veut pas que la maitresse Madame Etchebest (E) puisse intercepter leur communication. Ayant déjà des connaissances en cryptographie (oui c'est un génie ce petit) il propose à Arthur d'échanger une clef avec Béatrice pour communiquer de manière chiffrée. Le plan est infaillible seulement si l'institutrice n'intercepte pas la transmission de la clef. Le protocole d'échange fonctionne ainsi :

Arthur (A) génère une suite de $n$ bits aléatoires $b_1...b_n$. Pour chaque bit $b_i$ Arthur
- choisie une base de transmission $t_i$ entre $B_+ = \{\ket{0}, \ket{1}\}$ et $B_\times = \{\frac{\ket{0} + \ket{1}}{\sqrt{2}}, \frac{\ket{0} - \ket{1}}{\sqrt{2}}\} = \{\ket{+}, \ket{-}\}$.
- envoie un qubit encodé dans la base $t_i$ suivant ce tableau
<center>
<img src="img/tableau.png" width=200/>
</center>

Ensuite Béatrice (B), pour chaque qubit réceptionné $\ket{q_i}$
- choisie une base de réception $r_i \in \{B_+, B_\times \}$
- effectue une mesure dans la base de réception $r_i$, et obtient les bits $b'_i$

Dans un deuxième temps, Béatrice (B) transmet classiquement la liste des bases de réception $[r_1,...,r_n]$ à Arthur (A). Arthur dit donc à Béatrice quelles sont les bases $r_i$ et $t_i$ qui correspondent et gardent les qubits concernés pour créer la clef. 

En troisième étape Arthur (A) et Béatrice (B) dévoilent publiquement une partie de la clef. Ils comparents les $b_i$ et les $b'_i$. Si les bits sont identiques, la clef n'a pas été compromise et ils peuvent l'utiliser avec sécurité. Sinon, il y a eu un problème de transmission ou la maitresse Etchebest (E) a écouté cette transmission. 

La sécurité de la transmission repose sur le théorème de non-clonage. Il garantit qu'en cas d'écoute l'institutrice force le qubit à se projeter sur une base. Donc si l'institutrice devine la base choisi par Arthur avec une probabilité de 50%, alors 25% des bits de la clef finale seront erronés. Si Arthur et Béatrice sacrifient $m$ qubits dans la clef finale pour vérifier une éventuelle écoute de la maitresse alors ils peuvent détecter cette écoute avec une probabilité de $1 - (3/4)^m$.


Vous allez donc l'aider, lui et ses amis, à mettre se plan redoutable en pratique (parce que vous êtes un parent cool).

---
### Arthur (A) s'occupe de la transmission

Pour rappel, le protocole de transmission de la clef commence ainsi

Arthur (A) génère une suite de $n$ bits aléatoires $b_1...b_n$. Pour chaque bit $b_i$ Arthur
- choisie une base de transmission $t_i$ entre $B_+ = \{\ket{0}, \ket{1}\}$ et $B_\times = \{\frac{\ket{0} + \ket{1}}{\sqrt{2}}, \frac{\ket{0} - \ket{1}}{\sqrt{2}}\} = \{\ket{+}, \ket{-}\}$.
- envoie un qubit encodé dans la base $t_i$ suivant ce tableau
<center>
<img src="img/tableau.png" width=200/>
</center>


L'encodage dans la base $B_\times$ peut se faire en utilisant une porte H.

#### **Question 3.1**

*Implémenter une routine quantique qui prend en entrée le vecteur de n bits b, le vecteur qui précise les n bases de transmission t, et qui encode cela dans un état quantique suivant le tableau ci-dessus*

#### **Question 3.2**
*Testez votre fonction d'encodage sur l'exemple ci-dessous. Les états que vous devriez obtenir en sortie avec plot_histogram sont 0010, 0110, 1010, 1110*

In [80]:
n = 4
# Bits classiques
b = [0,1,1,1]
# Bases de transmission (0: B+, 1: Bx)
t = [0,0,1,1]


---
### Béatrice (B) réceptionne le message
Pour rappel, Béatrice (B), pour chaque qubit réceptionné $\ket{q_i}$
- choisie une base de réception $r_i \in \{B_+, B_\times \}$
- effectue une mesure dans la base de réception $r_i$, et obtient les bits $b'_i$

Effectuer une mesure dans la base $B_+$ correspond à une mesure basique, celle déjà implémentée dans Qiskit. Quant à la mesure dans la base $B_\times$, elle requiert d'appliquer une porte H avant une mesure basique.

#### **Question 3.3**
*Implémenter une routine quantique qui prends en entrée un vecteur de bases de réception r et qui effectue les mesures dans les bonnes bases*

---

### La maitresse Etchebest (E) écoute les élèves

En réalité, entre l'émission des qubits par Arthur (A) et leur réception par Béatrice (B) la maitresse Etchebest (E) peut intercepter des brides du transfert. Concrétement, considérons le vecteur $f$ tel que si $f_i = 0$ cela signifie qu'elle n'a pas écouté la transmission du qubit $i$ et si $f_i =1$ alors c'est qu'elle a écouté. Lorsqu'elle écoute la transmission à un moment $i$, elle effectue une mesure du qubit $i$. Elle a la liberté de choisir de le mesurer dans la base $B_+$ ou dans la base $B_\times$, ce choix est modélisé par le vecteur $e$, avec $e_i = 0$ si elle choisi la base $B_+$ à l'instant $i$, $e_i = 1$ si elle choisi plutôt la base $B_\times$.

#### **Question 3.4**
*Implémenter la routine quantique qui prend en entrée les vecteurs f et e et qui correspond aux écoutes de l'institutrice*

---

### La création de la clef

Maintenant Arthur possède une suite de bits $b$ et Béatrice $b'$. Ils vont d'abord comparer la listes des bases choisies, $t$ et $r$, et garder uniquement les bits envoyés et mesurés dans la même base.

#### **Question 3.5**
*Implémenter une fonction (purement classique) qui prend en entrée les vecteur t et r et qui garde que les indices pour lesquels les bases sont identiques*

Ils vont ensuite partager publiquement $m$ de leurs bits respectifs parmis ceux conservés. Si parmis eux certains ne sont pas identiques ils peuvent être sûres que la maitresse a espionné leur transmission, sinon ils peuvent affirmer avec une probabilité de $1 - (3/4)^m$ que ce n'est pas le cas. Pour faire simple ici, ils vont partager les $m$ premiers bits de la clef.

#### **Question 3.6**

*Implémenter une fonction booléene (purement classique) qui prend en entrée la valeur m et les vecteurs t, r, b et b', et qui renvoie si la transmission a été compromise*

---

### Le circuit complet

Il est désormais temps de mettre en place la mission. Votre ordinateur quantique peut utiliser tout ce que vous lui avez enseigné pour permettre à Arthur et Béatrice de communiquer secrétement en cours. Le circuit complet de l'opération se décompose de la manière suivante :
1) Arthur (A) transmet les $n$ qubits encodés à Béatrice (B)
2) L'institutrice Etchebest (E) écoute la conversation
3) Béatrice (B) réceptionne les qubits et les mesures
4) Arthur et Béatrice vérifient si la clé est compromise en partageant $m$ bits

#### **Question 3.7**
*Implémenter une routine quantique qui prend en entrée n, m, b, t, r, f et e et qui renvoie le circuit quantique correspondant aux étapes 1 à 3 (comprises)*

#### **Question 3.8**
*Tester le circuit quantique et l'obtention du résultat de compromission dans le cas suivant, où il n'y a pas d'écoute de la part de l'instritutrice.
Pour cela, vous pourrez effectuer l'expériences 100 fois et voir avec quelle fréquence on obtient un résultat 
de compromission de la clef (en théorie 0)*

In [81]:
n = 10
m = 4
b = [0,1,1,0,1,0,1,1,1,0]
t = [0,1,1,1,0,1,0,1,0,0]
r = [0,1,0,1,1,1,0,1,0,0]
f = [0,0,0,0,0,0,0,0,0,0]
e = [0,0,0,0,0,0,0,0,0,0]

#### **Question 3.9**
*Tester le circuit quantique dans ce cas où la maitresse écoute partiellement la communication et l'obtention du 
résultat de compromission de la clef sur plusieurs essais comme pour la question 3.8*

In [82]:
n = 10
m = 4
b = [0,1,1,0,1,0,1,1,1,0]
t = [0,1,1,1,0,1,0,1,0,0]
r = [0,1,0,1,1,1,0,1,0,0]
f = [1,1,0,1,0,0,0,0,0,0]
e = [1,1,0,1,0,0,1,0,1,0]

Comme $m=4$, la probabilité de détecter un espion est de $1 - (3/4)^m \approx 0.7$, vous devez donc théoriquement obtenir une fréquence supérieure à cette valeur.